### PR TITLE
Integration with Micrometer for greater ZIO monitoring capabilities

### LINKED ISSUE #24: CollectorRegistry (prometheus)
Many libraries (tapir, sttp, etc) integrate directly with prometheus `CollectorRegistry` so we need a way to combine and publish ZIO-core metrics with `CollectorRegistry` metrics.

Possibly  https://github.com/zio/zio-metrics-connectors/issues/20 solves this problem.
Micrometer provides `PrometheusMeterRegistry`.

Until  there is no solution to work with `CollectorRegistry`, I think, the `zio-metrics` should be undepricated. We need it rewritten with Module Pattern 2.0 :)

### LINKED ISSUE #34: Follow Prometheus Exposition format in Prometheus Encoder
When I connect the prometheus published metrics from my service which includes DefaultJvmMetrics to [Telegraf](https://www.influxdata.com/time-series-platform/telegraf/), I see several parsing errors as follows:
```
2023-03-09T16:31:20Z E! [inputs.prometheus] Error in plugin: error reading metrics for http://localhost:9000/metrics: reading text format failed: text format parsing error in line 7: second TYPE line for metric name "jvm_memory_pool_bytes_max", or TYPE reported after samples
```
I see the error for various metric names when using DefaultJvmMetrics, including but not limited to: `jvm_memory_pool_bytes_init`, `jvm_memory_bytes_init`, `jvm_memory_pool_bytes_max`, `jvm_memory_bytes_max`, `jvm_gc_collection_seconds_sum`, `jvm_memory_pool_bytes_used`, `jvm_memory_pool_bytes_max`.

I notice that this is because the Prometheus Encoder implementation does not exactly follow the exposition format (https://prometheus.io/docs/instrumenting/exposition_formats/#comments-help-text-and-type-information), specifically, the rule that "Only one TYPE line may exist for a given metric name."

The metrics with the same metric name but different labels should be combined together with exactly one TYPE line.

The actual output from DefaultJvmMetrics with prometheus publisher:
```
# TYPE jvm_gc_collection_seconds_count gauge
# HELP jvm_gc_collection_seconds_count
jvm_gc_collection_seconds_count{gc="G1 Young Generation",} 4.0 1678469178428
...
...
# TYPE jvm_gc_collection_seconds_count gauge
# HELP jvm_gc_collection_seconds_count
jvm_gc_collection_seconds_count{gc="G1 Old Generation",} 0.0 1678469178428
```

The expected output:
```
# TYPE jvm_gc_collection_seconds_count gauge
# HELP jvm_gc_collection_seconds_count
jvm_gc_collection_seconds_count{gc="G1 Old Generation",} 0.0 1678469178428
jvm_gc_collection_seconds_count{gc="G1 Young Generation",} 4.0 1678469178428
```

### LINKED ISSUE #41: Create an metrics connector to send metrics to OpenTelemetry
## Context
I've seen quite a lot of clients pair up OpenTelemetry collectors in their Kubernetes clusters as a way to aggregate and send both traces and metrics to end providers like New Relic, Datadog, etc.

## Ask 
It would be great to have an opentelemetry metrics integration built directly into the project. OpenTelemetry provides a [Metrics API](https://opentelemetry.io/docs/specs/otel/metrics/api) that we can integrate with using:
* [gRPC](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto)
* REST + JSON

Additional References:
* https://opentelemetry.io/docs/specs/otel/protocol/otlp
* https://opentelemetry.io/docs/specs/otel/protocol/otlp/#json-protobuf-encoding

Note: It seems like an implementation of #20 may be able to support this as [Micrometer has support for OpenTelemetry](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryMeterRegistry.java)

### PR DESCRIPTION
**Pull request by Grryum**
I am pleased to announce the successful integration of ZIO 2 Metrics with Micrometer, solving the problem of the lack of any universal connector of the ZIO ecosystem and the rest of the Java world in the context of metrics. This merge request introduces a seamless integration module that connects ZIO Metrics with Micrometer, allowing to translate all captured metric data from ZIO application to any of famous observability systems via common Micrometer facade.

## Benefits:

1. Standardized Metrics Collection: Micrometer is a widely adopted and standardized metrics library in the Java ecosystem. By integrating ZIO Metrics with Micrometer, we align our metrics collection approach with the industry-standard practices followed by the Java community. This ensures compatibility and interoperability with a wide range of existing monitoring systems, tools, and frameworks.

2. Universal facade (SLF4J but for metrics): Micrometer offers extensive support for various monitoring backends, including popular solutions like Prometheus, Datadog, Graphite, InfluxDB, and more. By integrating ZIO Metrics with Micrometer, we gain access to Micrometer's ecosystem of exporters, dashboards, and visualization tools. This allows us to leverage the existing integrations and take advantage of the rich features provided by Micrometer for metrics visualization, alerting, and analysis.

3. Easy Integration with Java Libraries: The integration of ZIO Metrics with Micrometer simplifies the integration of ZIO applications with other Java libraries and frameworks. Many Java libraries and frameworks already provide out-of-the-box integration with Micrometer, allowing us to seamlessly capture metrics from both ZIO and non-ZIO components of our applications. This facilitates a unified observability approach across the entire Java application stack.

4. Stay on Trend with Java Metrics Development: Micrometer is a popular and actively maintained metrics library in the Java ecosystem. By integrating ZIO Metrics with Micrometer, we align ourselves with the latest developments and updates in the Java metrics space. This ensures that our monitoring infrastructure stays up-to-date and can adapt to new features, formats, and backend updates introduced by Micrometer. No need to manually encode metrics for different end systems, such as Prometheus, Datadog, etc (like in other modules of zio-metric-connectors). Instead, the current format of the end system can be maintained centrally using Micrometer.

5. Real-time Metrics Updates: ZIO Metrics updates are immediately reflected in Micrometer, ensuring accurate monitoring.

## Technical details:

- The integration utilizes a low-level integration approach directly from the ZIO Metrics core client listener to Micrometer.
- The integration provides real-time synchronization between ZIO Metrics and Micrometer for accurate and timely monitoring. The module ensures immediate updates to Micrometer whenever metric changes occur within the ZIO application.
- Module depends on ZIO core and Micrometer core only.

## New document module was also added.

Please review this merge request, and I welcome any feedback or suggestions for further improvements.

Closes #24  #20
Probably closes #41

P.S. I am not sure that I did all requirements for proper document addition, I will be very grateful if you tell me how I can check the final documentation build.

### CODE REVIEW

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  private lazy val gauges: ConcurrentHashMap[MetricKey[Gauge], AtomicDouble] =
+    new ConcurrentHashMap[MetricKey[Gauge], AtomicDouble]()
+
+  override def updateHistogram(
+    key: MetricKey[MetricKeyType.Histogram],
+    value: Double,
+  )(implicit unsafe: Unsafe,
+  ): Unit =
+    DistributionSummary
+      .builder(key.name)
```
Comment (xlvp) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:38]: It seems that the name of the metric should conform to the [convention](https://micrometer.io/docs/concepts#_naming_meters). 
See PrometheusEncoder#encodeMetric

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      .tags(micrometerTags(key.tags).asJava)
+      .serviceLevelObjectives(key.keyType.boundaries.values.filter(_ > 0): _*) // micrometer prohibits <= 0 slo values
+      .register(meterRegistry)
+      .record(value)
+
+  override def updateGauge(key: MetricKey[Gauge], value: Double)(implicit unsafe: Unsafe): Unit =
+    gauges
```
Comment (zaxxel) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:33]: Only if absent?

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      .computeIfAbsent(
```
Comment (zaxxel) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:34]: Only if absent?

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        key,
+        _ =>
+          meterRegistry.gauge(
+            key.name,
+            micrometerTags(key.tags).asJava,
```
Comment (xlvp) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala]: Also, you should extract the description and remove it from the tags

File: docs/metrics/micrometer-connector.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+---
+id: micrometer-connector
+title: Micrometer Connector
+---
+
+ZIO ZMX also has an integration with [Micrometer](https://micrometer.io/), a powerful metrics instrumentation library.
```
Comment (vagroz) [inline docs/metrics/micrometer-connector.md]: ZIO ZMX is a separated and deprecated project https://github.com/zio/zio-zmx

Looks like this project is based on that one

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala (hunk 1 above)
Comment (Grryum) [reply micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:38]: I thought about it but decided to let user set basic name of metric key that would be thrown further. After the metric is processed by the vendor-specific Micrometer backend, the metric key name is automatically formatted according to the conventions of that backend. For example, when using PrometheusMeterRegistry, counters have "_total" appended to the metric name, and duration metrics have "_seconds" appended, etc. 

Please let me know if you have any further concerns or suggestions regarding this approach.

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala (hunk 4 above)
Comment (Grryum) [reply micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala]: I would appreciate clarification on what you mean by "description" the context of tag.
I have a method for building micrometer Tag from zio MetricLabel:

```
  private[micrometer] def micrometerTags(zioMetricTags: Iterable[MetricLabel]): Iterable[Tag] =
    zioMetricTags.map(metricTag => Tag.of(metricTag.key, metricTag.value))
```
There is no description in zio MetricLabel

File: docs/metrics/micrometer-connector.md (hunk 5 above)
Comment (Grryum) [reply docs/metrics/micrometer-connector.md]: sure! you are right, ZMX mention should be totally removed from docs

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      .distributionStatisticExpiry(key.keyType.maxAge)
+      .publishPercentiles(key.keyType.quantiles: _*)
+      .percentilePrecision(3) // should this be added to config?
+      .register(meterRegistry)
+      .record(value)
+
+  override def updateCounter(key: MetricKey[Counter], value: Double)(implicit unsafe: Unsafe): Unit =
+    meterRegistry
+      .counter(key.name, micrometerTags(key.tags).asJava)
+      .increment(value)
```
Comment (xlvp) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:79]: I will try to explain the above using the example of a counter

```suggestion
    {
    // add config: zio.metrics.connectors.MetricsConfig field to the constructor

    val name = key.name.replaceAll("-", ".").trim
    val description = key.tags.find(_.key == config.descriptionKey).map(_.value)
    val tagsWithoutDescription = key.tags.filterNot(_.key == config.descriptionKey).map(micrometerTags)

    Counter
      .builder(name)
      .description(description.orNull)
      .tags(tagsWithoutDescription.asJava)
      .register(registry)
      .increment(value)
  }
```

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala (hunk 4 above)
Comment (Grryum) [reply micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala]: I made a PR to ZIO Metrics core https://github.com/zio/zio/pull/8159. Let's wait for review. If it be accepted, I would be able to fix description extraction

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      .builder(key.name)
+      .tags(
+        (micrometerTags(key.tags) ++ Iterable(Tag.of("error", key.keyType.error.toString))).asJava,
+      )
+      .distributionStatisticBufferLength(key.keyType.maxSize)
+      .distributionStatisticExpiry(key.keyType.maxAge)
+      .publishPercentiles(key.keyType.quantiles: _*)
+      .percentilePrecision(config.summaryPercentileDigitsOfPrecision)
+      .register(meterRegistry)
+      .record(value)
```
Comment (petoalbert) [inline micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:74]: I think we could improve the performance of this implementation if we would not "build" the metrics each time. Similar to what you did with the gauges, we could have a map of `MetricKey.Summary` to `DistributionSummary` instances that are already registered with micrometer, and the same for all metric types.  I think that should be at least 2 times faster. What do you think?

**Comment by petoalbert** [comment]
This looks great @Grryum . I left one comment related to performance.

File: micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala (hunk 7 above)
Comment (Grryum) [reply micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala:74]: @petoalbert Thank you for your suggestion to improve the performance of the implementation.
In regards to your suggestion, I had actually considered a similar approach during the initial development. However, after conducting a investigation, I concluded that the performance improvement would be minimal due to Micrometer's internal metric storage mechanism.

Micrometer internally utilizes a HashMap with a key of Meter.Id to store all the registered metrics. When we create a metric using the MeterBuilder, it just provides a more flexible way to configure and create the initial Meter. However, subsequent calls to the same MeterBuilder for the same metric identity will return the already created Meter instance from the internal storage https://www.javadoc.io/doc/io.micrometer/micrometer-core/latest/io/micrometer/core/instrument/DistributionSummary.Builder.html#register(io.micrometer.core.instrument.MeterRegistry).

The optimization you suggested, which involves maintaining a map of MetricKey to Micrometer Meters instances, would primarily eliminate the need for additional allocations during the fluent Micrometer Builder creations. However, we would have the double concurrent Meters storages in our app.

I appreciate your time you took to consider performance enhancements. If you have any additional suggestions or concerns, please feel free to share them. I am open to exploring alternative approaches to further improve the performance and efficiency of the ZIO Metric integration.

**Review by petoalbert — APPROVED** [review]

**Pull request merged**